core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -223,21 +223,20 @@ class LogSegment(val log: FileRecords,
     timeIndex.resize(timeIndex.maxIndexSize)
     var validBytes = 0
     var lastIndexEntry = 0
-    maxTimestampSoFar = Record.NO_TIMESTAMP
+    maxTimestampSoFar = RecordBatch.NO_TIMESTAMP
     try {
-      for (entry <- log.shallowEntries(maxMessageSize).asScala) {
-        val record = entry.record
-        record.ensureValid()
+      for (entry <- log.batches(maxMessageSize).asScala) {
+        entry.ensureValid()
 
         // The max timestamp should have been put in the outer message, so we don't need to iterate over the inner messages.
-        if (record.timestamp > maxTimestampSoFar) {
-          maxTimestampSoFar = record.timestamp
-          offsetOfMaxTimestamp = entry.offset
+        if (entry.maxTimestamp > maxTimestampSoFar) {
+          maxTimestampSoFar = entry.maxTimestamp
+          offsetOfMaxTimestamp = entry.lastOffset
         }
 
         // Build offset index
         if(validBytes - lastIndexEntry > indexIntervalBytes) {
-          val startOffset = entry.firstOffset
+          val startOffset = entry.baseOffset
```

Comment:
I think the invariant we are trying to maintain is that the offsets added to the index are strictly increasing, so using the base offset in place of the first offset seems sufficient and avoids the decompression needed to find the actual first offset.